File: 백준/Silver/5464. 주차장/주차장.kt
package org.example

import java.util.PriorityQueue
import java.util.ArrayDeque

fun main() {
    val br = System.`in`.bufferedReader()

    // 입력 처리
    val (n, m) = br.readLine().split(" ").map{it.toInt()}
    val parkingPrice = IntArray(n){br.readLine().toInt()} // 주차 공간 별 요금
    val carWeights = IntArray(m+1) // 차량 무게 (번호가 1~n이어서 0은 안 쓰고 둠)
    for(i in 1..m){
        carWeights[i] = br.readLine().toInt()
    }

    // 주차 상태 관리
    val availableSpaces = PriorityQueue<Int>() // 빈 공간 번호 관리
    for(i in 0 until n){
        availableSpaces.add(i)
    }
    val parking = IntArray(n){-1} // 주차 공간 상태 (-1: 빈 상태)
    val waitingQueue = ArrayDeque<Int>() // 대기 차량
    var totalPrice = 0

    // 차량 입출력 처리
    repeat(2*m){
        val order = br.readLine().toInt()

        if(order >0 ){ // 차량 입장
            if(availableSpaces.isNotEmpty()){
                // 빈 공간이 있는 경우
                val space = availableSpaces.poll()
                parking[space] = order
                totalPrice += parkingPrice[space] * carWeights[order]
            }else {
                // 빈 공간이 없는 경우 대기
                waitingQueue.add(order)
            }
        }else{
            // 차량 퇴장
            val car = -order
            for(i in 0 until n){
                if(parking[i] == car){
                    parking[i] = -1 // 공간 비우기
                    availableSpaces.add(i) // 빈 공간 추가
                    break
                }
            }
            // 대기 차량 처리
            if (waitingQueue.isNotEmpty() && availableSpaces.isNotEmpty()){
                val space = availableSpaces.poll()
                val nextCar= waitingQueue.poll()
                parking[space] = nextCar
                totalPrice += parkingPrice[space] * carWeights[nextCar]
            }
        }
    }
    // 결과 출력
    println(totalPrice)
}

File: 백준/Silver/5464. 주차장/주차장.py
import sys
from collections import deque
import heapq

input = sys.stdin.readline

# 입력 처리
n, m = map(int, input().split())
parking_price = [int(input()) for _ in range(n)]
m_weight = [0] + [int(input()) for _ in range(m)]  # 차량 번호 1부터 시작

# 주차 상태 관리
available_spaces = list(range(n))  # 빈 공간을 관리하는 힙 (0부터 n-1까지 공간 번호)
heapq.heapify(available_spaces)
parking = [-1] * n  # 주차 공간 상태 (-1: 비어 있음)
waiting_queue = deque()  # 대기 차량
total_price = 0

# 차량 입출입 처리
for _ in range(2 * m):
    order = int(input())

    if order > 0:  # 차량 입장
        if available_spaces:  # 빈 공간이 있는 경우
            space = heapq.heappop(available_spaces)
            parking[space] = order
            total_price += parking_price[space] * m_weight[order]
        else:  # 빈 공간이 없는 경우 대기
            waiting_queue.append(order)

    else:  # 차량 퇴장
        car = -order
        for i in range(n):
            if parking[i] == car:
                parking[i] = -1  # 공간 비우기
                heapq.heappush(available_spaces, i)  # 빈 공간 추가
                break

        # 대기 차량 처리
        if waiting_queue and available_spaces:
            space = heapq.heappop(available_spaces)
            next_car = waiting_queue.popleft()
            parking[space] = next_car
            total_price += parking_price[space] * m_weight[next_car]

# 결과 출력
print(total_price)
